### PR TITLE
 Implements Startup/Setup Functionality for vSphere CSI Driver

### PR DESCRIPTION
**Pull request by dvonthenen**
**What this PR does / why we need it**:
This implements the functionality in the vSphere CSI driver to Startup/Setup the service to:
- consume a vsphere.conf file and create it's corresponding Config object
- takes the previous Config object to generates corresponding vSphere connections via ConnectionManager
- make use of a new common pkg `kubernetes` to create a k8s client and secret listener for cred manager. More info below

Build/Deploy related updates:
- adds YAML files to conveniently deploy vSphere CSI image to a kubernetes cluster
- provides starter/WIP documentation to deploy vSphere CSI 
- updates the Makefile to build the docker CSI image and uploads to said container to the configured registry

CCM updates:
- makes use of the new InformerManager class that both CCM and CSI now use to generate a secret listener for cred manager.

Common packages updates:
- removed overlapping roles and role-bindings in the CCM's YAMLs
- adds a new `kubernetes` pkg which provides helper functions to create a kubernetes client based on various configuration options and implements a wrapper for all Informers
- the `New` func to create a ConnectionManager was consolidated to generate the specific configuration based on the vsphere.conf Config object.

**Which issue this PR fixes**: NA

**Testing performed**:
`make test` passes all tests
Tested vSphere Cloud Controller Manager on Kubernetes 1.11.3
Tested vSphere CSI on Kubernetes 1.11.3
- verified the k8s client is created successfully
- verified the connection manager was able to connect to all configured vcenters
- k8s secretlistener was used

**Special notes for your reviewer**:
Some common packages were refactored to be easily consumed by CCM and CSI. These changes do not impact the deployment, configuration files, or user experience for either projects.

The updates to the Makefile includes functionality to upload packages to the configured container registry. I am not sure if anything needs to be configured on the Google Container Registry to allow the vsphere-csi image to be loaded to it.

### CODE REVIEW

File: pkg/csi/service/fcd/controller.go [hunk 1]
```diff
@@ -17,19 +17,61 @@ limitations under the License.
...
+
+	//testing
+	/*
+		err = connMgr.Verify()
+		if err == nil {
+			log.Infoln("connMgr.Verify() Succeeded")
+		} else {
+			log.Errorln("connMgr.Verify() Failed. Err:", err)
+		}
+	*/
```
Comment (frapposelli) [inline pkg/csi/service/fcd/controller.go]: leftover?

Comment (dvonthenen) [reply pkg/csi/service/fcd/controller.go]: I left it intentional for @codenrhoden as an example to test. He can remove when he continues to build out the provider.

Comment (frapposelli) [reply pkg/csi/service/fcd/controller.go]: sounds good, maybe make the comment more explicit there?

**Comment by dvonthenen** [comment]
@akutz it looks like the `pull-cloud-provider-vsphere-test` job is failing from the gcr upload failing. Looks like my concern about this not working automagically was correct.

**Comment by akutz** [comment]
Hi @dvonthenen,

> Looks like my concern about this not working automagically was correct.

To what are you referring?

> it looks like the `pull-cloud-provider-vsphere-test` job is failing from the gcr upload failing

Hi @figo, you set up this test, right? Can you please assist @dvonthenen with this? Thank you.

**Comment by akutz** [comment]
Hi @dvonthenen,

FWIW, the [build log](https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/cloud-provider-vsphere/103/pull-cloud-provider-vsphere-test/165/) appears to indicate `make test` is failing. Is this related to the GCR upload? The build log doesn't seem to mention the GCR upload at all as far as I am able to discern.

Again, I think @figo will need to assist you since I was not involved in any of the Prow configuration for this project, and there has been no sync up on it. I will ask @figo to use today's test meeting sync to review the Prow work he has done.

**Comment by dvonthenen** [comment]
Yea, I just noticed it. I retested locally and it is fine. Going to run it one more time.

**Comment by akutz** [comment]
Hi @dvonthenen,

It seems to be failing here ([log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/cloud-provider-vsphere/103/pull-cloud-provider-vsphere-test/166/build-log.txt)):

```shell
I1127 17:46:31.705] --- FAIL: TestDiscoverNodeByName (0.01s)
I1127 17:46:31.705]     nodemanager_test.go:141: Failed DiscoverNode: NotAuthenticated
I1127 17:46:31.705]     nodemanager_test.go:145: Failed: nodeNameMap should be a length of 1
I1127 17:46:31.705]     nodemanager_test.go:148: Failed: nodeUUIDMap should be a length of  1
```

**Comment by akutz** [comment]
My previous comment was for build 166. Build 167 ([log]()) just failed here:

**Comment by dvonthenen** [comment]
@akutz just capturing the convo here... 166 happens sometimes when I do a rebase and squash. still need to figure out why it happens once in a blue moon

167 seems to be the real error and i believe is failing with bind on the RPC server

**Comment by dvonthenen** [comment]
@akutz Added a retry on the go test for the client connecting to the grpc server. Looks like it fixed it.

Looking for feedback on the PR now.

CC: @frapposelli @codenrhoden @dougm

File: pkg/csi/service/fcd/controller.go (hunk 1 above)
Comment (dvonthenen) [reply pkg/csi/service/fcd/controller.go]: Done!

File: pkg/common/kubernetes/informers.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		client:          client,
+		stopCh:          signals.SetupSignalHandler(),
+		informerFactory: informers.NewSharedInformerFactory(*client, noResyncPeriodFunc()),
+	}
+}
+
+// GetSecretListener creates a lister to use
+func (im *InformerManager) GetSecretListener() listerv1.SecretLister {
+	if im.secretInformer == nil {
+		im.secretInformer = im.informerFactory.Core().V1().Secrets()
```
Comment (dougm) [inline pkg/common/kubernetes/informers.go]: Is this thread safe or should there be a mutex around the `secretInformer` field?

File: pkg/common/kubernetes/informers.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	return im.secretInformer.Lister()
+}
+
+// AddNodeListener hooks up add, update, delete callbacks
+func (im *InformerManager) AddNodeListener(add, remove func(obj interface{}), update func(oldObj, newObj interface{})) {
+	if im.nodeInformer == nil {
+		im.nodeInformer = im.informerFactory.Core().V1().Nodes().Informer()
```
Comment (dougm) [inline pkg/common/kubernetes/informers.go]: Is this thread safe or should there be a mutex around the `nodeInformer` field?

File: pkg/common/connectionmanager/connectionmanager.go [hunk 4]
```diff
@@ -111,6 +124,19 @@ func (cm *ConnectionManager) Logout() {
 	}
 }
 
+func (cm *ConnectionManager) Verify() error {
+	for vcServer := range cm.VsphereInstanceMap {
+		err := cm.Connect(context.Background(), vcServer)
```
Comment (dougm) [inline pkg/common/connectionmanager/connectionmanager.go]: Consider passing `Context` as a param to `Verify()`, so there's an option to use a timeout for testing these connections.

File: pkg/common/connectionmanager/connectionmanager.go [hunk 5]
```diff
@@ -111,6 +124,19 @@ func (cm *ConnectionManager) Logout() {
...
+		if err == nil {
+			glog.V(3).Infof("vCenter connect %s succeeeded.", vcServer)
```
Comment (dougm) [inline pkg/common/connectionmanager/connectionmanager.go]: typo

Comment (dvonthenen) [reply pkg/common/connectionmanager/connectionmanager.go]: Will fix that.

File: pkg/common/connectionmanager/connectionmanager.go (hunk 4 above)
Comment (dvonthenen) [reply pkg/common/connectionmanager/connectionmanager.go]: Good suggestion. Will provide both.

File: pkg/common/kubernetes/informers.go (hunk 2 above)
Comment (dvonthenen) [reply pkg/common/kubernetes/informers.go]: Yes, I just checked the SharedInformerFactor that informerFactory is derived from has a mutex on it.

File: pkg/common/kubernetes/informers.go (hunk 3 above)
Comment (dvonthenen) [reply pkg/common/kubernetes/informers.go]: Same as above.

**Comment by codenrhoden** [comment]
If the GCR stuff is working, this PR will close #76

**Pull request merged**